hw3/src/GrowList.java

```diff
@@ -0,0 +1,73 @@
+import java.util.*;
+//  GrowList is a mutable list that only gets longer.
+public class GrowList <E> {
+
+    private Map<Integer,E> values;
+
+    private GrowList() { values = new HashMap<Integer,E>();}
+
+    // add to the end of the list
+    public GrowList add(E o) {
+        GrowList<E> result = new GrowList<E>();
+        result.values = values;
+        result.values.put(size(), o);
+        return result;
+    }
+
+    // number of values in list
+    public int size() { return values.size(); }
+
+    // get ith value in list
+    public E get(int i) {
+        if (! inRange(i)) throw new IndexOutOfBoundsException("GrowList.get");
+        return values.get(i);
+    }
+
+    // update ith value in list;  return previous value or null if none
+    public E set(int i, E o) {
+        if (! inRange(i)) throw new IndexOutOfBoundsException("GrowList.set");
+        // producer
+        GrowList<E> result = new GrowList<E>();
+        for (Integer j = 0; j < size(); j++) {
+            if (j == i) {
+                result.add(o);
+            } else {
+                result.add(values.get(j));
+            }
+        }
+        values = result.values;
+        // observer
+        return values.get(i);
```

Comment:
Posted a question about whether the mutator needs to be split into two methods as opposed to having a method be both a producer and an observer: https://piazza.com/class/jl84sbpvd70246?cid=29

At the very least, I think the observer needs to be placed in a separate method, even if the producer calls the observer.